### DISCUSSION TITLE
Energy Minimization

### DISCUSSION THREAD

**Discussion by tortijazz** [netket/netket#900] [Q&A] [answered]
Hi everybody,

I was trying to reproduce the one-dimensional Heisenberg model example reported in the article "NetKet: A machine learning toolkit for many-body quantum systems", but instead using nk.operator.Heisenberg I used nk.operator.GraphOperator to compare the results using both methods (that in principle, should give the same results). In order to define a custom Hamiltonian and minimize the energy, I read the /legacy/CustomHamiltonian notebook as a guided example. The problem is when I try to find the ground state (of the custom Hamiltonian), the minimizer gets stuck at -7.00 energy value. My question is, Anybody knows why this happens? I thought the problem was the minimizer, but I tried out all available by Netket and the results don't change much.
<img width="1680" alt="Screenshot 2021-08-30 at 15 19 55" src="https://user-images.githubusercontent.com/46231608/131400460-d5bedad6-8f45-454f-bc6a-735654f0eb99.png">
<img width="1680" alt="Screenshot 2021-08-30 at 15 22 20" src="https://user-images.githubusercontent.com/46231608/131400467-f9b293fb-a13f-47c9-987e-cbef4e5ef0d5.png">

**Comment by PhilipVinc** [comment]
If you post a copy-pasteable script it would make life easier for people to help you.

I think a possible problem Is that your ansatz is real-positive, and I don't think the ground state you're looking for is too.
You might try using complex-valued parameters to get a complex output, for example.

**Reply by tortijazz** [reply]
This is the code:

```
import netket as nk
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import json

%matplotlib inline

# Define custom graph
L = 20

edges = []
for i in range(L):
    edges.append([i, (i+1)%L, 1])
    
# Define the netket graph object
g = nk.graph.Graph(edges=edges)
# Definir el grafo con Chain

# Printing out the graph information
print('This graph has', g.n_nodes, 'sites')

# Pauli matrices
sigmaz = [[1, 0], [0, -1]]
sigmax = [[0, 1], [1, 0]]
sigmay = [[0, -1j], [1j, 0]]

# Bond operator
interaction = np.kron(sigmaz, sigmaz) + np.kron(sigmax, sigmax) + np.kron(sigmay, sigmay)

bond_operator = [interaction.tolist()]

bond_color = [1]

hi = nk.hilbert.Spin(s = 1/2, total_sz = 0, N = g.n_nodes)

ha = nk.operator.GraphOperator(hi, bond_ops = bond_operator, bond_ops_colors = bond_color, graph = g)

res = nk.exact.lanczos_ed(ha, k = 4, compute_eigenvectors=False)
print("Exact ground state energy = {0:.2f}".format(res[0]))

ma = nk.models.RBM(alpha = 1, use_visible_bias = False, use_hidden_bias = True, dtype = float,)

# Metropolis Local Sampling
sa = nk.sampler.MetropolisExchange(hilbert = hi, graph = g, n_chains = 16)
print(sa)

# Optimizer
op = nk.optimizer.Sgd(learning_rate = 0.1)
sr = nk.optimizer.SR(diag_shift = 0.1)

# Variational state
vs = nk.vqs.MCState(sa, ma, n_samples = 10000, n_discard_per_chain = 100)

# Variational Monte Carlo driver
gs = nk.VMC(ha, op, variational_state = vs, preconditioner = sr)

# Print parameter structure
print(f"Number of variational parameters: {gs.state.n_parameters}")

gs.run(n_iter = 3000, out = "test2")
```

The ansatz is `nk.models.RBM`, which the same in the article.

**Reply by tortijazz** [reply]
Thank you, Filip. Your suggestion was right.

**Comment by fabienalet** [comment, marked as answer]
The problem is indeed what @PhilipVinc suggested: real-valued RBM. 
`ma = nk.models.RBM(alpha = 1, use_visible_bias = False, use_hidden_bias = True, dtype = complex)`
should do the trick.

**Reply by PhilipVinc** [reply]
Yes, in the old version of netket we were using complex parameters by default, but switched to real in the last release.
real-valued RBM only give you real output. That's why it is not working.

**Reply by tortijazz** [reply]
It works! Thank you both.
